dockers/docker-orchagent/ipinip_ping.py

```diff
@@ -0,0 +1,197 @@
+            tunnel = self.config_db.get_keys('TUNNEL')[0]
+        except IndexError:
+            logger.log_warning('PEER_SWITCH or TUNNEL table'
+                               'not found in config DB')
+            return None, None
+
+        try:
+            peer_loopback_ip = self.config_db.get_entry(
+                                    'PEER_SWITCH', peer_switch
+                                    )['address_ipv4']
+            self_loopback_ip = self.config_db.get_entry(
+                                    'TUNNEL', tunnel
+                                    )['dst_ip']
+        except KeyError as e:
+            logger.log_warning(
+                'PEER_SWITCH or TUNNEL table missing data, '
+                'could not find key {}'
+                .format(e)
+            )
+            return None, None
+
+        return self_loopback_ip, peer_loopback_ip
+
+    def get_inner_pkt_type(self, packet):
+        """
+        Get the type of an inner encapsulated packet
+
+        Returns:
+            (str)  'v4' if the inner packet is IPv4
+            (str)  'v6' if the inner packet is IPv6
+            (bool) False if `packet` is not an IPinIP packet
+        """
+        if packet.haslayer(IP):
+            # Determine inner packet type based on IP protocol number
+            # The outer packet type should always be IPv4
+            if packet[IP].proto == 4:
+                return IP
+            elif packet[IP].proto == 41:
+                return IPv6
+        return False
+
+    def listen_for_ipinip(self):
+        """
+        Listens for IPinIP packets that are trapped to CPU
+
+        These packets may be trapped if there is no neighbor info for the
+        inner packet destination IP in the hardware.
+        """
+
+        def _ping_inner_dst(packet):
+            """
+            Pings the inner destination IP for an IPinIP packet
+
+            Args:
+                packet: The IPinIP packet received
+            """
+            inner_packet_type = self.get_inner_pkt_type(packet)
+            if inner_packet_type and packet[IP].dst == self_ip:
+                cmds = ['ping', '-c1', '-W1']
+                if inner_packet_type == IPv6:
+                    cmds.append('-6')
+                dst_ip = packet[IP][inner_packet_type].dst
+                cmds.append(dst_ip)
+                logger.log_info("Running command '{}'".format(' '.join(cmds)))
+                subprocess.run(cmds, stdout=subprocess.DEVNULL)
+
+        self_ip, peer_ip = self.get_ipinip_tunnel_addrs()
+        if self_ip is None or peer_ip is None:
+            logger.log_error('Could not get IPinIP tunnel addresses from '
+                             'config DB, exiting...')
+            return None
+
+        packet_filter = 'host {} and host {}'.format(self_ip, peer_ip)
+        logger.log_notice('Starting IPinIP listener for IPs {} and {}'
+                          .format(self_ip, peer_ip))
+        while True:
+            sniff(
```

Comment:
could you check the cpu utilization of this? especially, if have a lots of packet trapped to Lpbk0 IP?

Reply:
The maximum rate that scapy can send packets from the PTF is about 170 KB/s. Sending IPinIP packets at this rate causes utilization to hover around ~2%, never exceeding ~2.6%
```
admin@str2-7050cx3-acs-08:~$ ps ax | grep ipinip_ping | grep -v grep
  91073 pts/0    S      0:03 python3 /usr/bin/ipinip_ping.py
admin@str2-7050cx3-acs-08:~$ top | grep 91073
  91073 root      20   0   38704  28308   9376 S   2.3   0.2   0:03.65 python3
  91073 root      20   0   38960  28364   9376 S   2.0   0.2   0:03.71 python3
  91073 root      20   0   38960  28432   9376 S   1.7   0.2   0:03.76 python3
  91073 root      20   0   38960  28508   9376 S   2.6   0.2   0:03.84 python3
  91073 root      20   0   38960  28568   9376 S   1.7   0.2   0:03.89 python3
  91073 root      20   0   38960  28628   9376 S   2.0   0.2   0:03.95 python3
  91073 root      20   0   39216  28692   9376 S   2.0   0.2   0:04.01 python3
  91073 root      20   0   39216  28772   9376 S   2.0   0.2   0:04.07 python3
  91073 root      20   0   39216  28840   9376 S   2.0   0.2   0:04.13 python3
  91073 root      20   0   39348  28904   9376 S   2.3   0.2   0:04.20 python3
```